### PR TITLE
Add support for openings in panels

### LINKED ISSUE #57: Add support for openings in Panels
#### Description:
creating and reading panels do not support openings in the panels. Add support for this.

### PR DESCRIPTION
**Pull request by Arne-Martensen**
### Issues addressed by this PR

 Closes #57

 
In the previous version the pull and push of panel did not include openings inside of those. The pulled/pushed panel was only able to transfer panel without openings. Additionally, the push was only able to include a single panel. 
In the current version it should be possible to push and pull any number or panels with an arbitrary number of openings between RFEM and GH.

 ### Test files

[File](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?id=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F02%5FPull%20Request%2FBHoM%2FRFEM%5FToolkit%2FRFEM%5FToolkit%5Fread%20and%20write%20panels%20with%20openings%2Egh&parent=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F02%5FPull%20Request%2FBHoM%2FRFEM%5FToolkit)

 ### Changelog

-Push/Pull of openings included in panels in RFEM_Toolkit.
-Push/Pull of arbitrary number of panels.

### CODE REVIEW

File: RFEM_Adapter/CRUD/Create/Panel.cs [hunk 1]
```diff
@@ -61,19 +61,44 @@ private bool CreateCollection(IEnumerable<Panel> panels)
...
                     foreach (Edge e in panelList[i].ExternalEdges)
                     {
+                        
                         //create rfem nodes, i.e. bhom points - NOTE: RFEM will remove the coincident points itself leaving jumps in node numbering ! 1,2,4,6,8,10,...
-                        Line edgeAsLine = e.Curve as Line;
 
-                        rf.Node rfNode1 = new rf.Node
+                        rf.Node rfNode1 = new rf.Node();
+
+                        if (e.Curve.GetType().Name.Equals("Polyline"))
```
Comment (IsakNaslundBh) [inline RFEM_Adapter/CRUD/Create/Panel.cs:69]: Think the above works, but you can simply do this instead, which imo reads a bit easier, and should run a bit faster.
```suggestion
                        if (e.Curve is Polyline)
```

File: RFEM_Adapter/CRUD/Create/Panel.cs [hunk 2]
```diff
@@ -101,9 +125,86 @@ private bool CreateCollection(IEnumerable<Panel> panels)
...
+                        {
+                            openingId = modelData.GetLastObjectNo(rf.ModelObjectType.OpeningObject);
+                            List<string> openingOutlineNodeList = new List<string>();
+                            
+                            //Defining Nodes
+                            foreach (Edge e in openingList[o].Edges)
+                            {
+                                
+                                rf.Node rfNode = new rf.Node();
+                                if (e.Curve.GetType().Name.Equals("Polyline"))
```
Comment (IsakNaslundBh) [inline RFEM_Adapter/CRUD/Create/Panel.cs:146]: As above

```suggestion
                                if (e.Curve is Polyline)
```

**Review by IsakNaslundBh — COMMENTED** [review]
Think it generally looks great.

There is some potential to refactor a few things to their own methods (for example a method that accepts a list of edges, creates them and returns a list of node ids. this method could then be used both for external edges of panels as well as openings) but this is not something that I would let hold up this PR.

I left some quick suggestions for type checking of the inner curve as well. Will not put this PR as a request for change though to not block in case someone is testing.
Would later maybe be good to have checks for other curve types than Lines and polylines, but think this also could be handled separately, as this PR does not make things worse (only improvement on that front).

I do not have access to RFEM, so am unfortunately not able to test this. @MartinHenriksen @staintono would any of you guys be able to assist with the functionality testing of this?

In general, looks great!

**Comment by staintono** [comment]
With the provided test file the surfaces with opening can be pushed and pulled to/from RFEM. However using my own model in RFEM I get the following error: 
_

> 1. This component failed to run properly.
> - Error: Input string was not in a correct format.
> - Occured in  BH.Adapter.RFEM.RFEMAdapter.GetIdListFromString(String idsAsString)
>      called from BH.Adapter.RFEM.RFEMAdapter.ReadPanels(List`1 ids)
> - Are you sure you have the correct type of inputs? Check their description for more details.

_
This is using the following RFEM [file](https://burohappold-my.sharepoint.com/:f:/g/personal/oliver_stainton_burohappold_com1/EsevGBgJ3M5CrbmA0MAepMwBV0Wqm8DLJ0-6CCpuhnJprQ?e=dnW6Jr) 

@Arne-Martensen @MartinHenriksen is this directly related to this commit, or a general panel issue?

**Comment by Arne-Martensen** [comment]
@staintono I don't know what exactly could be the reason but I do have a suspicion, currently the push and pull only works with panels defined by lines and polylines, could that be the reason? Could you maybe share the model so I can have a look?

**Comment by Arne-Martensen** [comment]
> 



> With the provided test file the surfaces with opening can be pushed and pulled to/from RFEM. However using my own model in RFEM I get the following error: _
> 
> > 1. This component failed to run properly.
> > 
> > 
> > * Error: Input string was not in a correct format.
> > * Occured in  BH.Adapter.RFEM.RFEMAdapter.GetIdListFromString(String idsAsString)
> >   called from BH.Adapter.RFEM.RFEMAdapter.ReadPanels(List`1 ids)
> > * Are you sure you have the correct type of inputs? Check their description for more details.
> 
> _ This is using the following RFEM [file](https://burohappold-my.sharepoint.com/:f:/g/personal/oliver_stainton_burohappold_com1/EsevGBgJ3M5CrbmA0MAepMwBV0Wqm8DLJ0-6CCpuhnJprQ?e=dnW6Jr)
> 
> @Arne-Martensen @MartinHenriksen is this directly related to this commit, or a general panel issue?
> 
> UPDATE: I have taken the openings out of the RFEM model and it still throws the same error, so this seems to be more an issue with the panels rather than the openings.

@staintono I've found the issue. There was a bug in my commit. The issue was that the algorithm worked fine when all panels had an opening but ran into problems as soon as there were panels present without openings. I fixed it now.

**Review by staintono — APPROVED** [review]
Panels with openings can now successfully be pulled along with the relevant openings. Works as expected with rectangular openings. Does not work with circular openings. This will be raised as a new issue.

**Comment by IsakNaslundBh** [comment]
@BHoMBot check compliance

**Review by staintono — APPROVED** [review]
Works as expected, see previous review.

**Comment by Arne-Martensen** [comment]
@BHoMBot

**Comment by Arne-Martensen** [comment]
@BHoMBot check compliance

**Comment by MartinHenriksen** [comment]
@BHoMBot check required

**Comment by MartinHenriksen** [comment]
@BHoMBot check ready-to-merge

**Pull request merged**